### PR TITLE
Start a high-level architecture document for Wasmtime

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit cleands up some existing documentation by removing a number
of "noop README files" and starting a high-level overview of the
architecture of Wasmtime. I've placed this documentation under the
contributing section of the book since it seems most useful for possible
contributors.

I've surely left some things out in this pass, and am happy to add more!

### CODE REVIEW

File: docs/contributing-architecture.md [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+the instruction, trapping if the signatures mismatch. This is implemented in
+Wasmtime as an integer comparison, and the comparison happens on a
+`VMSharedSignatureIndex` value. This index is an intern'd representation of a
+function type.
+
+The scope of interning for `VMSharedSignatureIndex` happens at the
+`wasmtime::Engine` level, which while not discussed in detail here can be
+thought of as a layer above `wasmtime::Store` intended to be shared across
+threads. Modules are compiled into an `Engine`. Insertion of a `Module` into
+n `Engine` will assign a `VMSharedSignatureIndex` to all of the types found
```
Comment (bjorn3) [inline docs/contributing-architecture.md:229]: *an

File: docs/contributing-architecture.md [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+3. The `InstanceHandle` corresponding to this instance is allocated. How this
+   is allocated depends on the strategy (malloc for on-demand, slab allocation
+   for pooling). In addition to initialization of the fields of `InstanceHandle`
+   this also initializes all the fields of the `VMContext` for this handle
+   (which as mentioned above is adjacent to the `InstanceHandle` allocation
+   after it in memory). This does not process any data segments, element
+   segments, or the `start` function at this time.
+
+4. At this point the `InstanceHandle` is persisted within the `Store`. This is
+   the "point of no return" of the handle must be retained for the next
```
Comment (bjorn3) [inline docs/contributing-architecture.md:274]: I can't parse this sentence.

File: docs/contributing-architecture.md [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+know that linear memories are contiguous this results in a 6GB reservation per
+memory because the guard region after one memory is the guard region before the
+next.
+
+## Tables and `externref`
+
+WebAssembly tables contain reference types, currently either `funcref` or
+`externref`. A `funcref` in Wasmtime is represented as `*mut
+VMCallerCheckedAnyfunc` and an `externref` is represented as `*mut
+VMExternRef`. Tables are consequently represented as vectors of pointers.
```
Comment (fitzgen) [inline docs/contributing-architecture.md:400]: An `externref` is actually `VMExternRef`, not `*mut VMExternRef`, and `VMExternRef` is a transparent newtype over `*mut VMExternData`.

**Review by fitzgen — APPROVED** [review]

File: docs/contributing-architecture.md [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+on is thought of as an internal dependency. We publish crates to crates.io but
+put very little effort into having a "nice" API for internal crates or worrying
+about breakage between versions of internal crates. This primarily means that
+all the other crates discussed here are considered internal dependencies of
+Wasmtime and don't show up in the public API of Wasmtime at all. To use some
+Cargo terminology, all the `wasmtime-*` crates that `wasmtime` depends on are
+"private" dependencies. While we strive to make the internal crates have proper
+safe/unsafe annotations for their APIs in reality almost everything they do is
+`unsafe` and much of it is not tagged as such. It's an eventual goal though to
+have `unsafe` be properly annotated for all of the internal crates of Wasmtime.
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply docs/contributing-architecture.md:28]: My goal with this was to document the state of reality. The reality is that every time I look at `wastime-runtime` I see more bugs of "oh well this isn't marked `unsafe` but it should be". I fully agree this isn't a great reality but I also don't want to lure anyone into a false sense of security with a false reality.

Perhaps I could reword this to try to indicate this better?

File: crates/cranelift/README.md [hunk 5]
```diff
@@ -1,4 +0,0 @@
-# `wasmtime-cranelift`
```
Comment (cfallin) [inline crates/cranelift/README.md:1]: A note on this and other `README.md` deletions: there was a recent PR #2987 that added READMEs where missing so that a downstream packager could satisfy distro requirements. I think we should probably keep these around, even if they are somewhat trivial one-sentence descriptions, as it's fairly low-cost. (If the concern is that the docs could distract from the "real" architecture doc, perhaps we could include a pointer?)

Comment (alexcrichton) [reply crates/cranelift/README.md:1]: From that PR it's not clear to me at least if the requirement was that `README.md` must be present or "the one in the source isn't present in `cargo package`". Do you know if it's one of those two?

I personally find no use for one-liner readmes that explain nothing beyond the title of the crate, and I personally find it to be a worse experience not seeing a README in the first place and getting my hopes up that one exists. It's similar to the feeling of a function's documentation not explaining anything beyond the name of the function.

I also find that these sorts of README.md files which aren't at the root level of a repository are perennially neglected. If we really want to add documentation to these crates I think we should do it in a centralized location (the book) and more technical details would go in the source itself (e.g. a large crate doc comment)

Comment (cfallin) [reply crates/cranelift/README.md:1]: Ah! No, I think you're right, the issue looks like it was actually a metadata problem, i.e. already existing README not included in the package. So if that's the case, and packagers don't see an issue, I'm totally happy to get rid of them. (cc @olivierlemasle to confirm?)

Comment (olivierlemasle) [reply crates/cranelift/README.md:1]: :+1: Yes, my objective was to include the existing READMEs, but no README is not an issue for me.

File: docs/contributing-architecture.md [hunk 6]
```diff
@@ -22,10 +22,15 @@ about breakage between versions of internal crates. This primarily means that
 all the other crates discussed here are considered internal dependencies of
 Wasmtime and don't show up in the public API of Wasmtime at all. To use some
 Cargo terminology, all the `wasmtime-*` crates that `wasmtime` depends on are
-"private" dependencies. While we strive to make the internal crates have proper
-safe/unsafe annotations for their APIs in reality almost everything they do is
-`unsafe` and much of it is not tagged as such. It's an eventual goal though to
-have `unsafe` be properly annotated for all of the internal crates of Wasmtime.
+"private" dependencies.
+
+Additionally at this tiem the safe/unsafe boundary between Wasmtime's internal
```
Comment (abrown) [inline docs/contributing-architecture.md:27]:
```suggestion
Additionally at this time the safe/unsafe boundary between Wasmtime's internal
```

**Comment by alexcrichton** [comment]
Ok I'm gonna go ahead and merge this. We can of course continue to iterate in-tree, though!
